src/core/qgsnetworkcontentfetcherregistry.cpp

```diff
@@ -0,0 +1,132 @@
+/***************************************************************************
+                       qgsnetworkcontentfetcherregistry.cpp
+                             -------------------
+    begin                : April, 2018
+    copyright            : (C) 2018 by Denis Rouzaud
+    email                : denis@opengis.ch
+
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsnetworkcontentfetcherregistry.h"
+
+#include "qgsapplication.h"
+#include "qgsnetworkcontentfetchertask.h"
+
+QgsNetworkContentFetcherRegistry::QgsNetworkContentFetcherRegistry()
+  : QObject()
+{
+}
+
+QgsNetworkContentFetcherRegistry::~QgsNetworkContentFetcherRegistry()
+{
+  QMap<QUrl, QgsFetchedContent *>::const_iterator it = mFileRegistry.constBegin();
+  for ( ; it != mFileRegistry.constEnd(); ++it )
+  {
+    it.value()->mFile->close();
+    delete it.value()->mFile;
+  }
+  mFileRegistry.clear();
+}
+
+const QgsFetchedContent *QgsNetworkContentFetcherRegistry::fetch( const QUrl &url, const FetchingMode &fetchingMode )
+{
+  if ( mFileRegistry.contains( url ) )
+  {
+    return mFileRegistry.value( url );
+  }
+
+  QgsFetchedContent *content = new QgsFetchedContent( nullptr, QgsFetchedContent::NotStarted );
+  QgsNetworkContentFetcherTask *fetcher = new QgsNetworkContentFetcherTask( url );
```

Comment:
This looks leaky -there's paths here where fetcher isn't added to the task manager. Can you use a unique ptr instead?